### ISSUE TITLE
Undue restrictions on adding types with System.Reflection.Metadata.Ecma335.MetadataBuilder

### ISSUE THREAD

**Issue by August-Alm** [dotnet/runtime#80360]
### Description

The API exposed by the MetadataBuilder class for emitting metadata for (sub)class definitions imposes restrictions that are contrary to the ECMA-335 CIL standard.

### Reproduction Steps

Consider the task of using the `MetadataBuilder` API for emitting the CIL corresponding to the following (contrived) C# code for a class with a subclass:

````CSharp
using System;

public class Outer
{
    static int Combine(int n, int m)
    {
        return n + m; 
    }

    private class Inner
    {
        public int Number { get; private set; }
        private Random _random;

        public Inner(int n)
        {
            _random = new Random();
            Number = Outer.Combine(n, _random.Next());
            if (Number == 0) throw new Exception("That's bad!");
        }
    }
    
    private Inner _inner;

    public Outer(int n)
    {
        _inner = new Inner(n);
    }

    public int Result => _inner.Number;
}
````

The key issue here is that the `Outer` constructor depends on the `Inner` class, which depends on the `Combine` method of the `Outer` class. A `MetadataBuilder` must thus define the `Combine`, `Inner.ctor` and `Outer.ctor` methods in that order (using the `AddMethodDefinition` method). More specifically, the reason is when emitting the `Call` instructions in the function bodies one must supply the `MethodDefinitionHandle`:s to the called functions; hence have those methods already defined. Similarly, the `Outer.ctor` needs not just that `Inner.ctor` has been defined, but that the `Inner` type has been defined (because it needs a handle to the `_inner` field, which for its definition needs a handle to the `Inner` type definition).

Now, to define the `Inner` subclass the `MetadataBuilder` API exposes a method `AddTypeDefinition`. One of its required arguments is a handle to the first method defined by the class. In this case, it must be the handle to the constructor. The `Inner` class will then, "own" all methods from that and until the first method handle owned by the next type definition (or, if no such exists, the last method defined in the module). But the same method must be used to define the `Outer` class, and this leads to an impossible dependency. For, if the the call to `AddTypeDefinition` that is supposed to define the `Outer` class is supplied the `Combine` handle (which must precede the `Inner.ctor` handle) or a handle preceding it, then the `Inner` class will own all methods succeeding the `Inner.ctor` handle, and one of those is the `Outer.ctor` handle, which should not belong to the subclass.

As another and related example of the problem,, consider the following situation:

````CSharp
using System.Runtime.InteropServices;

public class A
{
    [StructLayout(LayoutKind.Explicit, Pack = 1, Size = 10)]
    private struct SizeTen
    {
    }

    private SizeTen _sizeTen = <Marshalled raw data or defined by relative virtual address>;
}

````

The `AddTypeDefinition` invocation supposed to define the `SizeTen` struct is expected to supply not just a handle to the first method defined by the next type definition but also to the first field (since it does not own either fields or methods). The next field is `_sizeTen`, but it cannot be defined (using the `AddFieldDefinition` method of the `MetadataBuilder` class) without first having a handle to its, type. There is a mutually recursive dependency here that cannot be resolved by the exposed API.

### Expected behavior

If a piece of CIL metadata is conformant to the ECMA-335 specification, then the `MetadataBuilder` API should allow us to emit it. (Especially, I would argue, if the CIL in question is not a weird corner case but that which the Roslyn or F# compilers generate for quite common code patterns.)

### Actual behavior

The valid CIL code cannot be generated by the `MetadataBuilder` API.

### Known Workarounds

None.

### Configuration

.NET 6

### Other information

To summarise, it seems to me that the issue in question here is that ECMA 335 CIL imposes strict rules for how the various metadata tables must be sorted. An entry in the type definition table must for example contain an index saying which entry in the method definition table that is the first method that it owns. However, the `MetadataBuilder` API unduly extends these "(chrono)logical" orderings by imposing an ordering between elements _of different tables_. In the example above, we can imagine simultaneously adding the metadata for the type and the field, and end up with metadata conformant to section II.22.37 of the ECMA-335 specification.

**Comment by jkotas** [comment]
Roslyn generates the IL binaries using System.Reflection.Metadata APIs. Anything that can be written in C# can be produced using the System.Reflection.Metadata APIs.

Roslyn solves the problem with type definition cycles by pre-assigning tokens to the emitted types. It computes the order in which the types and methods are going to be emitted in first, and assigns them token with assumption that the first type added by `AddTypeDefinition` is going to get a token 1, second `AddTypeDefinition` is going to return a token 2, etc.

If it helps, https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs and https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/PEWriter/FullMetadataWriter.cs are the main Roslyn type that use System.Reflection.Metadata APIs to produce IL binaries. `DefinitionIndex` is the type responsible for pre-assigning tokens to the metadata entities.

**Comment by August-Alm** [comment]
Thank you for the clarification. I concede that there is a workaround: The  `MetadataTokens` part of the API essentially breaks the very encapsulation that the `MetadataBuilder` goes at such length to impose, and gives us methods like `TypeDefinitionHandle(int idx)` that allow us to index directly into metadata tables. This way one can supply any index whatsoever, one just has to make a mental note of how many definitions one has currently made and remember that "seven definitions from now, I have to make that definition that didn't yet exist when I needed it..." 

The situation is, I think, entirely analogous to the recently resolved issue (https://github.com/dotnet/runtime/issues/40546) with how the `InstructionEncoder` class handled switch branch instructions. It was entirely possible to emit the opcode and a companion list of "case" branch targets; one just had to manually calculate what those branch offsets into the as-yet-to-be-written code would be. However, such an API is, I think most would agree, poor library design. The adopted solution is to program with a `LabelHandle` abstraction instead of explicit IL offsets, and to supply a method `MarkLabel` that behind the scenes will fixate what offset the `LabelHandle` signifies. Functionality like this is what Roslyn builds on top of System.Reflection.Metadata API. But to say that the API supports it is, to me, a bit like saying that a BinaryWriter supports it.

**Comment by jkotas** [comment]
I agree that System.Reflection.Metadata does not completely abstract away metadata encoding details. It is a consequence of it being a very low-level and high-performance API. 

> The adopted solution is to program with a LabelHandle abstraction instead of explicit IL offsets, and to supply a method MarkLabel that behind the scenes will fixate what offset the LabelHandle signifies. Functionality like this is what Roslyn builds on top of System.Reflection.Metadata API. 

Roslyn does not use `System.Reflection.Metadata.Ecma335.InstructionEncoder` and `LabelHandle`s from `System.Reflection.Metadata`. It would introduce unnecessary overhead. Roslyn emits IL directly into blobs using its own encoder.

If you would like to see more convenience helper types like `InstructionEncoder` to make `System.Reflection.Metadata` easier to use, feel free to submit API proposal. However, `System.Reflection.Metadata` is never going to be as convenient to use as `System.Reflection.Emit`.

**Comment by August-Alm** [comment]
I didn’t mean that Roslyn uses the`InstructionEncoder`. I meant that (a) the latter class (since the recent Issue-fix) manages to keep `LabelHandle` encapsulated while still being managing to emit all CIL code one would like, by having the `MarkLabel` method, and (b), Roslyn does something comparable but for metadata such as type definitions, since the `Metadata` API on its own is obviously lacking in this regard.

Anyways, I’m closing this. I’ll think about if my issues can be turned into a feasible API proposal. (My vague proposal is that `TypeDefinitionHandle`:s and `MethodDefinitionHandle`:s, etc, should behave like `LabelHandle`:s.)
